typing/ctype.ml

```diff
@@ -4160,14 +4171,19 @@ let rec build_subtype env visited loops posi level t =
           (t, Unchanged)
       else
         (t, Unchanged)
-  | Tarrow(l, t1, t2, _) ->
+  | Tarrow((l,a,r) , t1, t2, _) ->
       if memq_warn t visited then (t, Unchanged) else
       let visited = t :: visited in
       let (t1', c1) = build_subtype env visited loops (not posi) level t1 in
       let (t2', c2) = build_subtype env visited loops posi level t2 in
-      let c = max c1 c2 in
-      (* FIXME update arrow modes *)
-      if c > Unchanged then (newty (Tarrow(l, t1', t2', Cok)), c)
+      let (a', c3) =
+        if level > 2 then build_submode (not posi) a else a, Unchanged
```

Comment:
`level > 2`?

Reply:
build_subtype only does its work down to some depth of expansions and object/variant types. It's completely ad-hoc how far to go and after some thought I decided that only replacing mode variables up until the first object/variant after and an expansion was reasonable. That corresponds to `2`.